app/src/main/java/net/gsantner/markor/format/ActionButtonBase.java

```diff
@@ -873,6 +880,37 @@ public static void moveLineSelectionBy1(final HighlightingEditor hlEditor, final
         }
     }
 
+    public static void duplicateLineSelection(final HighlightingEditor hlEditor) {
+        // Duplication is performed downwards, selection is moving alongside it.
+        final Editable text = hlEditor.getText();
+
+        final int[] sel = TextViewUtils.getSelection(hlEditor);
+        final int linesStart = TextViewUtils.getLineStart(text, sel[0]);
+        final int linesEnd = TextViewUtils.getLineEnd(text, sel[1]);
+
+        final CharSequence lines = text.subSequence(linesStart, linesEnd);
+
+        final int altStart = linesEnd + 1;
+        final int altEnd = TextViewUtils.getLineEnd(text, altStart);
+        final CharSequence altLine = text.subSequence(altStart, altEnd);
+
+        final int[] selStart = TextViewUtils.getLineOffsetFromIndex(text, sel[0]);
+        final int[] selEnd = TextViewUtils.getLineOffsetFromIndex(text, sel[1]);
+
+        hlEditor.withAutoFormatDisabled(() -> {
+            final String lines_final = String.format("%s\n", lines);
+            text.insert(altStart, lines_final);
+        });
+
+        final int sel_offset = selEnd[0] - selStart[0] + 1;
+        selStart[0] += sel_offset;
+        selEnd[0] += sel_offset;
+
+        hlEditor.setSelection(
```

Comment:
I don't think most of this is necessary. We only need to do this if we are modifying the selected lines. But we are not doing that here.

Reply:
I think this function can be reduced to something like

```
final int sel[] = TextViewUtils.getLineSelection(text);
final CharSequence lines = text.subSequence(sel[0], sel[1]);
hlEditor.withAutoFormatDisabled(() -> text.insert(sel[1] + 1, lines)
```

Please test :)

Reply:
Thanks for the quick feedback!
I just tried your suggestion, but there are a couple problems with it:
1) it is missing a newline at the end of the duplicated line(s)
2) it doesn't perserve the cursor column position

Please see later comment for more details on what I mean by point 2.